plugin/kvictoria/kvictoria.go

```diff
@@ -0,0 +1,243 @@
+package kvictoria
+
+import (
+	"errors"
+	"net"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	vm "github.com/VictoriaMetrics/metrics"
+	"github.com/twmb/franz-go/pkg/kerr"
+	"github.com/twmb/franz-go/pkg/kgo"
+)
+
+var (
+	// interface checks to ensure we implement the hooks properly
+	_ kgo.HookNewClient           = new(Metrics)
+	_ kgo.HookClientClosed        = new(Metrics)
+	_ kgo.HookBrokerConnect       = new(Metrics)
+	_ kgo.HookBrokerDisconnect    = new(Metrics)
+	_ kgo.HookBrokerWrite         = new(Metrics)
+	_ kgo.HookBrokerRead          = new(Metrics)
+	_ kgo.HookBrokerE2E           = new(Metrics)
+	_ kgo.HookBrokerThrottle      = new(Metrics)
+	_ kgo.HookGroupManageError    = new(Metrics)
+	_ kgo.HookProduceBatchWritten = new(Metrics)
+	_ kgo.HookFetchBatchRead      = new(Metrics)
+)
+
+// Metrics provides metrics using the [VictoriaMetrics/metrics] library.
+//
+// [VictoriaMetrics/metrics]: https://github.com/VictoriaMetrics/metrics
+type Metrics struct {
+	cfg cfg
+
+	sets map[string]*vm.Set
+}
+
+// NewMetrics returns a new Metrics that tracks metrics under the given namespace.
+//
+// You can pass options to configure the metrics reporting. See [Opt] for all existing options.
+func NewMetrics(namespace string, opts ...Opt) *Metrics {
+	return &Metrics{
+		cfg:  newCfg(namespace, opts...),
+		sets: make(map[string]*vm.Set),
+	}
+}
+
+// OnNewClient implements the [kgo.HookNewClient] interface for metrics gathering.
+// This method is meant to be called by the hook system and not by the user.
+func (m *Metrics) OnNewClient(client *kgo.Client) {
+	clientID := client.OptValue(kgo.ClientID).(string)
+
+	set, ok := m.sets[clientID]
+	if !ok {
+		set = vm.NewSet()
+		m.sets[clientID] = set
+	}
+
+	labels := map[string]string{"client_id": clientID}
+
+	set.GetOrCreateGauge(m.buildName("buffered_fetch_bytes_total", labels), func() float64 {
+		return float64(client.BufferedFetchBytes())
+	})
+	set.GetOrCreateGauge(m.buildName("buffered_fetch_records_total", labels), func() float64 {
+		return float64(client.BufferedFetchRecords())
+	})
+	set.GetOrCreateGauge(m.buildName("buffered_produce_bytes_total", labels), func() float64 {
+		return float64(client.BufferedProduceBytes())
+	})
+	set.GetOrCreateGauge(m.buildName("buffered_produce_records_total", labels), func() float64 {
+		return float64(client.BufferedProduceRecords())
+	})
+
+	vm.RegisterSet(set)
+}
+
+// OnClientClosed implements the [kgo.HookClientClosed] interface for metrics gathering.
+// This method is meant to be called by the hook system and not by the user.
+//
+// This will unregister all metrics that are scoped to the client id of the client provided.
+func (m *Metrics) OnClientClosed(client *kgo.Client) {
+	clientID := client.OptValue(kgo.ClientID).(string)
+
+	set, ok := m.sets[clientID]
```

Comment:
It's common for people to use the same clientID in many clients -- going off my comment above, I wonder if this package should document that a new `Metrics` should be created per client?

Otherwise, here, this may unregister a set that is used by multiple clients.

Reply:
Ok, rethinking about this more I think you're right. In that case I can also just drop the `sets` map and store a single set.